src/solve/slg/implementation/mod.rs

```diff
@@ -233,7 +233,176 @@ impl context::Environment<SlgContext> for Arc<Environment> {
     }
 }
 
-impl context::Substitution<SlgContext> for Substitution {}
+impl context::Substitution<SlgContext> for Substitution {
+    fn may_invalidate(&mut self, subst: &Substitution) -> bool {
+        self.parameters
+            .iter()
+            .zip(&subst.parameters)
+            .any(|(p1, p2)| MayInvalidate.aggregate_parameters(p1, p2))
+    }
+}
+
+// This is a struct in case we need to add state at any point like in AntiUnifier
+struct MayInvalidate;
+
+impl MayInvalidate {
+    fn aggregate_parameters(&mut self, p1: &Parameter, p2: &Parameter) -> bool {
```

Comment:
I think we should give the left- and right-hand-side more meaningful names. In particular:

- `p1` represents a potential new answer
- `p2` represents the aggregated answer thus far

right?

So maybe `new` and `current` as names (we could use same name no matter the type). Or else `new_` and `current_` as prefixes.